packages/core/src/tools/contextHistory.ts

```diff
@@ -3,83 +3,107 @@ import { relativeNow } from './timeUtils'
 import type { TimeoutId } from './utils'
 import { ONE_MINUTE } from './utils'
 
-interface PreviousContext<T> {
+const END_OF_TIMES = Infinity as RelativeTime
+
+export interface ContextHistoryEntry<T> {
   startTime: RelativeTime
   endTime: RelativeTime
   context: T
+  remove(): void
+  close(endTime: RelativeTime): void
 }
 
 export const CLEAR_OLD_CONTEXTS_INTERVAL = ONE_MINUTE
 
+/**
+ * Store and keep track of contexts spans. This whole class assumes that contexts are added in
+ * chronological order (i.e. all entries have an increasing start time).
+ */
 export class ContextHistory<Context> {
-  private current: Context | undefined
-  private currentStart: RelativeTime | undefined
-  private previousContexts: Array<PreviousContext<Context>> = []
+  private entries: Array<ContextHistoryEntry<Context>> = []
   private clearOldContextsInterval: TimeoutId
 
   constructor(private expireDelay: number) {
     this.clearOldContextsInterval = setInterval(() => this.clearOldContexts(), CLEAR_OLD_CONTEXTS_INTERVAL)
   }
 
-  find(startTime?: RelativeTime) {
-    if (
-      startTime === undefined ||
-      (this.current !== undefined && this.currentStart !== undefined && startTime >= this.currentStart)
-    ) {
-      return this.current
+  /**
+   * Add a context to the history associated with a start time. Returns a reference to this newly
+   * added entry that can be removed or closed.
+   */
+  add(context: Context, startTime: RelativeTime): ContextHistoryEntry<Context> {
+    const entry: ContextHistoryEntry<Context> = {
+      context,
+      startTime,
+      endTime: END_OF_TIMES,
+      remove: () => {
+        const index = this.entries.indexOf(entry)
+        if (index >= 0) {
+          this.entries.splice(index, 1)
+        }
+      },
+      close: (endTime: RelativeTime) => {
+        entry.endTime = endTime
+      },
     }
-    for (const previousContext of this.previousContexts) {
-      if (startTime > previousContext.endTime) {
+    this.entries.unshift(entry)
+    return entry
+  }
+
+  /**
+   * Return the latest context that was active during `startTime`, or the currently active context
+   * if no `startTime` is provided. This method assumes that entries are not overlapping.
+   */
+  find(startTime: RelativeTime = END_OF_TIMES): Context | undefined {
```

Comment:
I thought about it, but we almost always use `find()` with an optional argument (see [here](https://github.com/DataDog/browser-sdk/blob/6beff83922382ad407c0815341611218a8df9347/packages/rum-core/src/domain/viewContexts.ts#L11) [here](https://github.com/DataDog/browser-sdk/blob/6beff83922382ad407c0815341611218a8df9347/packages/rum-core/src/domain/urlContexts.ts#L17) [here](https://github.com/DataDog/browser-sdk/blob/6beff83922382ad407c0815341611218a8df9347/packages/core/src/domain/session/sessionManager.ts#L13)) so it would mean duplicate all those methods too, which is not ideal.